File: java/src/main/java/AuthorNameNormalizer.java
package util;

public class AuthorNameNormalizer {
    public String normalize(String name) {
    		throwWhenContainsTooManyCommas(name);
    		String[] parts = suffixless(name).trim().split(" ");
        if (isMononym(parts))
            return name;
        return lastName(parts) + ", " + firstName(parts) + 
        		middleInitials(parts) + suffix(name);
    }

	private void throwWhenContainsTooManyCommas(String name) {
		if (count(name, ',') > 1)
			throw new IllegalArgumentException();
		
	}
	private String suffixless(String name) {
		if(hasComma(name))
			return name.substring(0, name.indexOf(','));
		return name;
	}
	
	private boolean hasComma(String name) {
		return name.contains(",");
	}
	
	private String suffix(String name) {
		if(hasComma(name))
			return name.substring(name.indexOf(','));
		return "";
	}

    private String middleInitials(String[] parts) {
    		if (parts.length <= 2)
    			return "";
    		
    		String middleInitials = "";
    		for(int i = 1; i < parts.length - 1; i ++){
    			middleInitials += " " + initial(parts[i]);
    		}
    		
		return middleInitials;
	}

	private String initial(String name) {
		if(name.length() == 1) return name;
		return name.charAt(0) + ".";
	}

    
	private boolean isMononym(String[] parts) {
        return parts.length == 1;
    }

    private String firstName(String[] parts) {
        return parts[0];
    }

    private String lastName(String[] parts) {
        return parts[parts.length - 1];
    }

    // See http://stackoverflow.com/questions/275944/java-how-do-i-count-the-number-of-occurrences-of-a-char-in-a-string
    // ... if you need to convert to < Java 8
   /* private */ long count(String string, char c) {
        return string.chars().filter(ch -> ch == c).count();
    }
}

File: csharp/NameNormalizer/Prod/AuthorNameNormalizer.cs
using System;
using System.Collections.Generic;
using System.Linq;

namespace Prod
{
    public class AuthorNameNormalizer
    {
        public string Normalize(string name)
        {
            ThrowWhenContainsTooManyCommas(name);
            var parts = Suffixless(name).Trim().Split(' ');
            if (parts.Length == 1)
                return name;
            return $"{LastName(parts)}, {FirstName(parts)}{MiddleInitials(parts)}{Suffix(name)}";
        }

        private void ThrowWhenContainsTooManyCommas(string name)
        {
            if (name.Count(c => c == ',') > 1)
                throw new ArgumentException();
        }

        private string Suffix(string name)
        {
            var i = name.IndexOf(',');
            if (i == -1) return "";
            return name.Substring(i);
        }

        private string Suffixless(string name)
        {
            var i = name.IndexOf(',');
            if (i == -1) return name;
            return name.Substring(0, i);
        }

        private string MiddleInitials(string[] parts)
        {
            var middleNames = parts.Skip(1).Take(parts.Length - 2);
            return string.Join("", middleNames.Select(Initial));
        }

        private string Initial(string name)
        {
            var period = (name.Length > 1) ? "." : "";
            return $" {name[0]}{period}";
        }

        private string FirstName(string[] parts)
        {
            return parts[0];
        }

        private string LastName(string[] parts)
        {
            return parts[parts.Length - 1];
        }
        
    }
}
